docs/architecture/adr-008-ics20-implementation.md

```diff
@@ -0,0 +1,298 @@
+        fn is_bound(&self, port_id: PortId) -> bool;
+        /// get_transfer_account returns the ICS20 - transfers AccountId.
+        fn get_transfer_account(&self) -> AccountId;
+        /// get_port returns the portID for the transfer module.
+        fn get_port(&self) -> Result<PortId, Error>;
+    }
+
+    pub trait BankKeeper<AccountId> {
+        /// This function should enable sending ibc fungible tokens from one account to another
+        fn send_coins(&self, from: AccountId, to: AccountId, amt: Coin) -> Result<(), ICS20Error>;
+        /// This function to enable  minting ibc tokens in a module
+        fn mint_coins(&self, module: AccountId, amt: Coin) -> Result<(), ICS20Error>;
+        /// This function should enable burning of minted tokens
+        fn burn_coins(&self, module: AccountId, amt: Coin) -> Result<(), ICS20Error>;
+        /// This function should enable transfer of tokens from the ibc module to an account
+        fn send_coins_from_module_to_account(&self, module: AccountId, to: AccountId, amt: Coin) -> Result<(), ICS20Error>;
+        /// This function should enable transfer of tokens from an account to the ibc module
+        fn send_coins_from_account_to_module(&self, from: AccountId, module: AccountId, amt: Coin) -> Result<(), ICS20Error>;
+    }
+
+    pub trait AccountReader<AccountId> {
+        /// This function should return the account of the ibc module
+        fn get_module_account(&self) -> AccountId;
+        /// Returns the escrow account id for a port and channel combination
+        fn get_escrow_account(port_id: PortId, channel_id: ChannelId) -> AccountId;
+    }
+
+    pub trait ICS20Context: ICS20Keeper + ICS20Reader;
+```
+## Handling ICS20 Packets
+ICS20 messages are still a subset of channel packets so they should be handled as such
+
+The following handlers are recommended to be implemented in the ics20_fungible_token_transfer application in the ibc-rs crate
+
+These handlers will be executed in the module callbacks of any thirdparty IBCmodule that is implementing an ICS20 application on-chain
+```rust
+    pub enum ICS20Acknowledgement {
+        /// Equivalent to b"AQ=="
+        Success, 
+        /// Error Acknowledgement
+        Error(String)
+    }
+
+    impl From<GenericAcknoledgement> for ICS20Acknowledgement {
+        ...
+    }
+
+    impl Acknowledgement for ICS20Acknowledgement {
+        fn success() -> bool {
+            ...
+        }
+    }
+    /// Initiates ICS20 token transfer
+    /// Escrows tokens from sender and registers the send packet 
+    pub send_transfer<Ctx>(ctx: &Ctx, msg: MsgTransfer) -> Result<(), ICS20Error>
+        where Ctx: ICS20Context
+    {
+        /* 
+            ICS20 Application Logic
+            ...
+        */
+    }
+    
+    /// Handles incoming packets with ICS20 data
+    pub on_recv_packet<Ctx>(ctx: &Ctx, packet: Packet, data: MsgTransfer) -> Result<(), ICS20Error>
```

Comment:
We also need to define a `PacketData` type for this module with a deserialize impl, something like -> 
```rust
struct FungibleTokenPacketData {
  denomination: Denom
  amount: U256
  sender: AccountId
  receiver: AccountId
}
```